### PR TITLE
Feature(BWS/BWC) - XRP Support 

### PR DESCRIPTION
**Pull request by justinkook**
`bitcore.config.json`:

```
"XRP": {
        "mainnet": {
          "chainSource": "rpc",
          "provider": {
            "host": "s1.ripple.com",
            "protocol": "wss",
            "dataHost": "https://data.ripple.com"
          }
        },
        "testnet": {
          "chainSource": "rpc",
          "provider": {
            "host": "s.altnet.rippletest.net",
            "protocol": "wss"
          }
        }
      },
```

### CODE REVIEW

File: packages/bitcore-node/src/config.ts [hunk 1]
```diff
@@ -62,7 +62,7 @@ const Config = function(): ConfigType {
     dbPass: process.env.DB_PASS || '',
     numWorkers: cpus().length,
     chains: {},
-    modules: ['./bitcoin', './bitcoin-cash', './ethereum'],
```
Comment (micahriggan) [inline packages/bitcore-node/src/config.ts]: Revert this line and set it in your bitcore.config.json

File: packages/bitcore-wallet-service/src/config.ts [hunk 2]
```diff
@@ -56,6 +56,14 @@ module.exports = {
         url: 'https://api-eth.bitcore.io',
       },
     },
+    xrp: {
+      livenet: {
+        url: 'http://localhost:3000',
```
Comment (micahriggan) [inline packages/bitcore-wallet-service/src/config.ts]: Let's get a domain setup for this part

File: packages/bitcore-wallet-client/src/lib/credentials.ts [hunk 3]
```diff
@@ -176,6 +176,8 @@ export class Credentials {
         coin = '0';
       } else if (this.coin == 'eth') {
         coin = '60';
+      } else if (this.coin == 'xrp') {
+        coin = '144';
```
Comment (micahriggan) [inline packages/bitcore-wallet-client/src/lib/credentials.ts]: we should have a method for this in CWC, if not we should add that

File: packages/bitcore-wallet-client/src/lib/common/constants.ts [hunk 4]
```diff
@@ -52,11 +52,22 @@ export const Constants = {
...
+        maxDecimals: 6,
+        minDecimals: 2,
+      }
+    },
     usdc: {
       toSatoshis: 1e6,
       full: {
-        maxDecimals: 8,
-        minDecimals: 8,
+        maxDecimals: 6,
```
Comment (micahriggan) [inline packages/bitcore-wallet-client/src/lib/common/constants.ts]: ETH related change, Is this causing an issue in the app now?

Comment (justinkook) [reply packages/bitcore-wallet-client/src/lib/common/constants.ts]: reverted

File: packages/bitcore-node/src/config.ts (hunk 1 above)
Comment (justinkook) [reply packages/bitcore-node/src/config.ts]: reverted

File: packages/bitcore-wallet-client/src/lib/credentials.ts (hunk 3 above)
Comment (justinkook) [reply packages/bitcore-wallet-client/src/lib/credentials.ts]: https://github.com/bitpay/bitcore/pull/2640

File: packages/bitcore-wallet-client/src/lib/api.ts [hunk 5]
```diff
@@ -671,26 +672,25 @@ export class API extends EventEmitter {
...
-        const signedTxs = [];
-        for (let index = 0; index < signatures.length; index++) {
-          const signed = CWC.Transactions.applySignature({
-            chain,
-            tx: unsignedTxs[index],
-            signature: signatures[index]
-          });
-          signedTxs.push(signed);
+    const network = txp.network;
+    if (!Constants.UTXO_COINS.includes(chain.toLowerCase())) {
```
Comment (matiu) [inline packages/bitcore-wallet-client/src/lib/api.ts]: keeping the switch and just adding `case 'XRP':` before `case 'ETH'` could give us more flexibility for future coins.

File: packages/bitcore-wallet-client/src/lib/common/utils.ts [hunk 6]
```diff
@@ -369,13 +370,18 @@ export class Utils {
...
+        tag: destinationTag ? Number(destinationTag) : undefined
+      };
       const unsignedTxs = [];
       const isERC20 = tokenAddress && !payProUrl;
       const chain = isERC20 ? 'ERC20' : this.getChain(coin);
       for (let index = 0; index < recipients.length; index++) {
         const rawTx = Transactions.create({
           ...txp,
           ...recipients[index],
+          ...params,
```
Comment (matiu) [inline packages/bitcore-wallet-client/src/lib/common/utils.ts]: params and txp both have invoiceId.

if the objetive of params is to transform destinantionTag, maybe this could be refactored to:
```
   const rawTx = Transactions.create({
          ...txp,
          ...recipients[index],
          tag: destinationTag ? Number(destinationTag) : undefined,
          chain,
```
and avoid params.

File: packages/bitcore-wallet-service/src/lib/chain/eth/index.ts [hunk 7]
```diff
@@ -286,16 +286,14 @@ export class EthChain implements IChain {
...
   validateAddress(wallet, inaddr, opts) {
     const chain = 'ETH';
-    try {
-      Validation.validateAddress(
-        chain,
-        wallet.network,  // not really used for ETH. wallet.network is 'livenet/testnet/regtest' in wallet.
-        inaddr,
-      );
-    } catch (ex) {
+    const isValidAddress = Validation.validateAddress(
```
Comment (matiu) [inline packages/bitcore-wallet-service/src/lib/chain/eth/index.ts]: is this a bug? (ie validateAddress does not throw?). Can you split this on a different PR, and add a test. Thanks.

File: packages/bitcore-wallet-service/src/lib/chain/xrp/index.ts [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+        const convertedBalance = this.convertBitcoreBalance(balance, lockedSum);
+        server.storage.fetchAddresses(
+          server.walletId,
+          (err, addresses: IAddress[]) => {
+            if (err) return cb(err);
+            if (addresses.length > 0) {
+              const byAddress = [
+                {
+                  address: addresses[0].address,
+                  path: Constants.PATHS.SINGLE_ADDRESS,
```
Comment (matiu) [inline packages/bitcore-wallet-service/src/lib/chain/xrp/index.ts]: path: `addresses[0].path`

**Comment by cmgustavo** [comment]
I'm getting this error when trying to show "wallet-details" view. I'm not sure if I'm missing something:

```
ERR! /v3/wallets/?includeExtendedInfo=0&twoStep=0&serverMessageArray=1&r=31136 :404:404 - "<!DOCTYPE html>\n<html lang=\"en\">\n<head>\n<meta charset=\"utf-8\">\n<title>Error</title>\n</head>\n<body>\n<pre>Cannot POST //api/XRP/mainnet/wallet</pre>\n</body>\n</html>\n" 
<e4634f84-0984-491a-9818-cfe805f2ab74> ::1 2019-12-31T15:56:00.968Z "GET /bws/api/v3/wallets/?includeExtendedInfo=0&twoStep=0&serverMessageArray=1&r=31136" 404 217 537.113 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.88 Safari/537.36"  
```

File: packages/bitcore-wallet-service/src/config.ts (hunk 2 above)
Comment (justinkook) [reply packages/bitcore-wallet-service/src/config.ts]: pointed to `https://api-xrp.bitcore.io`

**Comment by justinkook** [comment]
> I'm getting this error when trying to show "wallet-details" view. I'm not sure if I'm missing something:
> 
> ```
> ERR! /v3/wallets/?includeExtendedInfo=0&twoStep=0&serverMessageArray=1&r=31136 :404:404 - "<!DOCTYPE html>\n<html lang=\"en\">\n<head>\n<meta charset=\"utf-8\">\n<title>Error</title>\n</head>\n<body>\n<pre>Cannot POST //api/XRP/mainnet/wallet</pre>\n</body>\n</html>\n" 
> <e4634f84-0984-491a-9818-cfe805f2ab74> ::1 2019-12-31T15:56:00.968Z "GET /bws/api/v3/wallets/?includeExtendedInfo=0&twoStep=0&serverMessageArray=1&r=31136" 404 217 537.113 "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.88 Safari/537.36"  
> ```

This happened because the URL was wrong in `config.ts`. It was pointed to `https://api-xrp.bitcore.io/` with an extra `/`. 

Fixed

File: packages/bitcore-wallet-service/src/lib/chain/xrp/index.ts (hunk 8 above)
Comment (justinkook) [reply packages/bitcore-wallet-service/src/lib/chain/xrp/index.ts]: fixed

File: packages/bitcore-wallet-client/src/lib/common/utils.ts (hunk 6 above)
Comment (justinkook) [reply packages/bitcore-wallet-client/src/lib/common/utils.ts]: done

File: packages/bitcore-wallet-service/src/lib/chain/eth/index.ts (hunk 7 above)
Comment (justinkook) [reply packages/bitcore-wallet-service/src/lib/chain/eth/index.ts]: Reverted to master and will make a separate PR

File: packages/bitcore-wallet-client/src/lib/api.ts (hunk 5 above)
Comment (justinkook) [reply packages/bitcore-wallet-client/src/lib/api.ts]: Changed back to switch statement
